### PR TITLE
Support custom Liquibase credentials

### LINKED ISSUE #31214: Dedicated DDL user for liquibase migrations to avoid DDL permissions for app datasource user
### Description

Similar to what was done for flyway in #24639.

It would be nice if it was possible to set a dedicated user for liquibase migrations so that only that user needs DDL permissions and the app datasource user only needs DML permissions.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by appiepollo14**
- Fixes https://github.com/quarkusio/quarkus/issues/31214

### CODE REVIEW

File: .mvn/extensions.xml [hunk 1]
```diff
@@ -19,4 +19,9 @@
         <artifactId>quarkus-project-develocity-extension</artifactId>
         <version>1.0.6</version>
     </extension>
+     <extension>
+         <groupId>org.apache.maven.extensions</groupId>
+         <artifactId>maven-build-cache-extension</artifactId>
+         <version>1.0.0</version>
+     </extension>
```
Comment (gsmet) [inline .mvn/extensions.xml:26]: Could you please revert this change?

**Review by gsmet — CHANGES_REQUESTED** [review]
This looks awesome but see my small gripe inline.

I will let others review this in details though.

File: .idea/runConfigurations/mvnDebug.xml [hunk 2]
```diff
@@ -4,10 +4,10 @@
     <option name="SERVER_MODE" value="false" />
     <option name="SHMEM_ADDRESS" />
     <option name="HOST" value="localhost" />
-    <option name="PORT" value="8000" />
+    <option name="PORT" value="5005" />
     <option name="AUTO_RESTART" value="false" />
     <RunnerSettings RunnerId="Debug">
-      <option name="DEBUG_PORT" value="8000" />
+      <option name="DEBUG_PORT" value="5005" />
```
Comment (gastaldi) [inline .idea/runConfigurations/mvnDebug.xml:10]: Please revert this change either

**Comment by gsmet** [comment]
As for PR title and commit message, I would go with something like:

> Allow to use specific datasource credentials for Liquibase
>
> Fixes #31214

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java [hunk 3]
```diff
@@ -31,9 +34,23 @@ public Liquibase createLiquibase() {
         try (ClassLoaderResourceAccessor resourceAccessor = new ClassLoaderResourceAccessor(
                 Thread.currentThread().getContextClassLoader())) {
 
-            Database database = DatabaseFactory.getInstance()
-                    .findCorrectDatabaseImplementation(new JdbcConnection(dataSource.getConnection()));
-            ;
+            Database database;
+
+            if (config.username.isPresent() && config.password.isPresent()) {
```
Comment (gastaldi) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java:39]: The Javadoc for `username` says: 
- `If no specific JDBC URL is configured, falls back to the datasource username if not specified.`
- 
And for `password`: 
- `If no specific JDBC URL is configured, falls back to the datasource password if not specified.`
 
This seems confusing as they are both ignored if one is not present.

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java [hunk 4]
```diff
@@ -31,9 +34,23 @@ public Liquibase createLiquibase() {
...
+                AgroalDataSource agroalDataSource = (AgroalDataSource) dataSource;
+                String jdbcUrl = agroalDataSource.getConfiguration().connectionPoolConfiguration()
+                        .connectionFactoryConfiguration().jdbcUrl();
+                Connection connection = DriverManager.getConnection(jdbcUrl, config.username.get(), config.password.get());
```
Comment (gastaldi) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java:43]: Can we use `datasource.getConnection(String username, String password)` here instead of using the Agroal API?

Comment (appiepollo14) [reply extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java:43]: No Than the error, in this https://github.com/quarkusio/quarkus/issues/31214#issuecomment-1991117911  discussion occurs.

File: integration-tests/liquibase/src/test/java/io/quarkus/it/liquibase/LiquibaseFunctionalityTest.java [hunk 5]
```diff
@@ -18,6 +18,20 @@ public void testLiquibaseQuarkusFunctionality() {
...
 
+    @Test
+    @DisplayName("Migrates a schema correctly using dedicated username and password from config properties")
+    public void testLiquibaseUsingDedicatedUsernameAndPassword() {
+        when().get("/liquibase/updateWithDedicatedUser").then().body(is(
+                "create-quarkus-table,insert-into-quarkus-table"));
+    }
+
+    @Test
+    @DisplayName("Checks the user")
```
Comment (appiepollo14) [inline integration-tests/liquibase/src/test/java/io/quarkus/it/liquibase/LiquibaseFunctionalityTest.java:29]: @gastaldi question. This test will only succeed if the `testLiquibaseUsingDedicatedUsernameAndPassword` is succeeded. So there is a dependency. Is that allowed?

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java [hunk 6]
```diff
@@ -31,9 +34,23 @@ public Liquibase createLiquibase() {
         try (ClassLoaderResourceAccessor resourceAccessor = new ClassLoaderResourceAccessor(
                 Thread.currentThread().getContextClassLoader())) {
 
-            Database database = DatabaseFactory.getInstance()
-                    .findCorrectDatabaseImplementation(new JdbcConnection(dataSource.getConnection()));
-            ;
+            Database database;
+
+            if (config.username.isPresent() && config.password.isPresent()) {
+                AgroalDataSource agroalDataSource = (AgroalDataSource) dataSource;
```
Comment (gastaldi) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/LiquibaseFactory.java:40]: Using `Datasource#unwrap` may be a better option than explicit casting
```suggestion
                AgroalDataSource agroalDataSource = dataSource.unwrap(AgroalDataSource.class);
```

File: integration-tests/liquibase/src/test/java/io/quarkus/it/liquibase/LiquibaseFunctionalityTest.java (hunk 5 above)
Comment (gastaldi) [reply integration-tests/liquibase/src/test/java/io/quarkus/it/liquibase/LiquibaseFunctionalityTest.java:29]: How about merging the checks with the test above?

**Comment by appiepollo14** [comment]
@gastaldi refactored. Can this be merged?

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java [hunk 7]
```diff
@@ -80,4 +80,16 @@ public class LiquibaseConfig {
      */
     public Optional<String> liquibaseTablespaceName = Optional.empty();
 
+    /**
+     * The username that Liquibase uses to connect to the database.
+     * If no username is configured, falls back to the datasource username.
```
Comment (gastaldi) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java:85]:
```suggestion
     * If no username is configured, falls back to the datasource username and password.
```

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java [hunk 8]
```diff
@@ -80,4 +80,16 @@ public class LiquibaseConfig {
...
+     */
+    public Optional<String> username = Optional.empty();
+
+    /**
+     * The password that Liquibase uses to connect to the database.
+     * If no password is configured, falls back to the datasource password.
```
Comment (gastaldi) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java:91]:
```suggestion
     * If no password is configured, falls back to the datasource password and username.
```

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java [hunk 9]
```diff
@@ -101,6 +101,20 @@ public static final LiquibaseDataSourceRuntimeConfig defaultConfig() {
     @ConfigItem
     public Optional<String> defaultSchemaName = Optional.empty();
 
+    /**
+     * The username that Liquibase uses to connect to the database.
+     * If no specific JDBC URL is configured, falls back to the datasource username if not specified.
```
Comment (gastaldi) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java:106]: This description is not really accurate, maybe use the same as above?

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java [hunk 10]
```diff
@@ -101,6 +101,20 @@ public static final LiquibaseDataSourceRuntimeConfig defaultConfig() {
...
+     */
+    @ConfigItem
+    public Optional<String> username = Optional.empty();
+
+    /**
+     * The password that Liquibase uses to connect to the database.
+     * If no specific JDBC URL is configured, falls back to the datasource password if not specified.
```
Comment (gastaldi) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseDataSourceRuntimeConfig.java:113]: Same here

**Review by gastaldi — CHANGES_REQUESTED** [review]
It looks good overall, just added a minor suggestion to make the config description clearer

File: extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java [hunk 11]
```diff
@@ -80,4 +80,16 @@ public class LiquibaseConfig {
...
 
+    /**
+     * The username that Liquibase uses to connect to the database.
+     * If no username is configured, falls back to the datasource username and password.
+     */
+    public Optional<String> username = Optional.empty();
+
+    /**
+     * The password that Liquibase uses to connect to the database.
+     * If no username is configured, falls back to the datasource username and password.
```
Comment (gastaldi) [inline extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java:91]:
```suggestion
     * If no password is configured, falls back to the datasource username and password.
```

Comment (appiepollo14) [reply extensions/liquibase/runtime/src/main/java/io/quarkus/liquibase/runtime/LiquibaseConfig.java]: sorry, too hastely.

**Review by gastaldi — APPROVED** [review]
LGTM, good job!

**Comment by gastaldi** [comment]
@appiepollo14 the build failures seem related to your change?

**Comment by appiepollo14** [comment]
@gastaldi It does. Will do some fixing.

**Comment by appiepollo14** [comment]
Disabled one IT on Windows as there is no Docker engine available. Added one extra class to the PMT test. Hopefully that will solve the missing dependency. 

@gastaldi is there a way to run the full CI on my fork?

**Comment by appiepollo14** [comment]
@gastaldi @gsmet this implementation is based on the Flyway implementation. To check the behavior of the extension with custom username / password, a second datasource (postgres) is introduced. This uses devservices to create a DB. The test now fail as, in opposition to the Flyway Integration tests, the PMT has no devservice PostGres database. Can be fixed maybe by reverting to a second datasource in H2, hopefully that has also a current_user default value for a table but should look into that. Other solution would be, to get rid of the PMT for the Liquibase Integration test. I prefer the second solution. Please approve / comment.

**Review by gsmet — APPROVED** [review]
I force pushed to rebase and fix the formatting error.

This patch looks very good and is definitely something we want in.

My only concern is that we should avoid adding a dependency to Hibernate ORM in the tests.
We should rather use Agroal directly and do a database query.

Making everything intertwined in Quarkus makes it very hard to do major changes and we already got hit several times by Hibernate being used all around.

Would you have the cycles to address this concern? It could be after we merged this PR as it's already very good.

Thanks!

**Comment by appiepollo14** [comment]
@gsmet I Will take on that challenge. However, there is still a failing Prod mode test. Please decide and reply to the following comment:

> @gastaldi @gsmet this implementation is based on the Flyway implementation. To check the behavior of the extension with custom username / password, a second datasource (postgres) is introduced. This uses devservices to create a DB. The test now fail as, in opposition to the Flyway Integration tests, the PMT has no devservice PostGres database. Can be fixed maybe by reverting to a second datasource in H2, hopefully that has also a current_user default value for a table but should look into that. Other solution would be, to get rid of the PMT for the Liquibase Integration test. I prefer the second solution. Please approve / comment.

**Comment by gastaldi** [comment]
What is PMT?

**Comment by gastaldi** [comment]
Maybe you can test with a single datasource, perform an INSERT in the test using the current user and read the table data to check if they match?

**Comment by appiepollo14** [comment]
@gastaldi Please advise

**Comment by gastaldi** [comment]
Why do you need a second datasource?

**Comment by appiepollo14** [comment]
Altered code, removed Postgres datasource in favor of H2 second database. Test should now work.

**Comment by appiepollo14** [comment]
@gsmet @gastaldi  removed hibernate and replaced it with agroal. It is partially working . I need te call, I can't figure out why, the endpoint for running Liquibase twice. After the second time I can call the endpoint: "created-by" and retrieve the inserted row from the database. After the first run, I get a database error stating the database is empty. 

This issue was introduced when removing hibernate in favor of agroal sql statement. So I believe it has to do something the agroal datasource. Might be settings?

Hope you guys have the solution so this PR can finally be merged!

File: integration-tests/liquibase/pom.xml [hunk 12]
```diff
@@ -1,7 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (gastaldi) [inline integration-tests/liquibase/pom.xml:2]: Revert the changes in this pom.xml

**Comment by appiepollo14** [comment]
@gastaldi @gsmet 
please guide me by providing an answer to the comment above. https://github.com/quarkusio/quarkus/pull/39361#issuecomment-2028832759

**Comment by appiepollo14** [comment]
Would love to see this pr merged. Help needed solving issue described here: https://github.com/quarkusio/quarkus/pull/39361#issuecomment-2028832759

Any ideas are welcome

**Comment by gsmet** [comment]
Hey there. Sorry I was busy on other things :).

I'll have a look now.

**Comment by appiepollo14** [comment]
@gsmet nice fix!!

**Comment by gsmet** [comment]
Merged! Thanks for your work and sorry for the delay!
